core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/NormalizeOrInExpressionRewriter.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner.iterative.rule;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMultimap;
+import com.google.common.collect.ImmutableSet;
+import io.trino.sql.tree.ComparisonExpression;
+import io.trino.sql.tree.Expression;
+import io.trino.sql.tree.ExpressionRewriter;
+import io.trino.sql.tree.ExpressionTreeRewriter;
+import io.trino.sql.tree.InListExpression;
+import io.trino.sql.tree.InPredicate;
+import io.trino.sql.tree.LogicalExpression;
+
+import java.util.Collection;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.sql.ExpressionUtils.and;
+import static io.trino.sql.ExpressionUtils.or;
+import static io.trino.sql.tree.ComparisonExpression.Operator.EQUAL;
+import static io.trino.sql.tree.LogicalExpression.Operator.AND;
+
+public final class NormalizeOrInExpressionRewriter
+{
+    public static Expression normalizeOrInExpression(Expression expression)
+    {
+        return ExpressionTreeRewriter.rewriteWith(new Visitor(), expression);
+    }
+
+    private NormalizeOrInExpressionRewriter() {}
+
+    private static class Visitor
+            extends ExpressionRewriter<Void>
+    {
+        @Override
+        public Expression rewriteLogicalExpression(LogicalExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            List<Expression> terms = node.getTerms().stream()
+                    .map(expression -> treeRewriter.rewrite(expression, context))
+                    .collect(toImmutableList());
+
+            if (node.getOperator() == AND) {
+                return and(terms);
+            }
+
+            ImmutableMultimap.Builder<Expression, Expression> expressionBuilder = ImmutableMultimap.builder();
+            for (Expression expression : terms) {
+                if (expression instanceof ComparisonExpression comparisonExpression && comparisonExpression.getOperator() == EQUAL) {
+                    expressionBuilder.put(comparisonExpression.getLeft(), comparisonExpression);
+                }
+                else if (expression instanceof InPredicate inPredicate && inPredicate.getValueList() instanceof InListExpression) {
+                    expressionBuilder.put(inPredicate.getValue(), inPredicate);
```

Comment:
why not to only values from `inPredicate` here?
```
expressionBuilder.put(inPredicate.getValue(), inPredicate.getValueList().getValues())
```
Same above:
```
 expressionBuilder.put(comparisonExpression.getLeft(), comparisonExpression.getRight());
 ```
 
 Then `mergeToInListExpression` would be trivial.

Reply:
The reason keep the original expression instead expanding the values is similar to the above, want to keep the relative position from inputs when we should not rewrite the expression. Take instance: the expression `I1 = 1 OR I2 IN (1,2,3) OR I3 = 3` should not be rewrite, but if expand the values, we will see `I2` has 3 values, but we do not know it's actual just comes from 1 inList, thus will change the expression to `I1 = 1 OR I3 = 3 OR I2 IN (1, 2, 3)`